org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/concurrent/RequestCancelIndicator.java

```diff
@@ -0,0 +1,45 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2019 TypeFox GmbH (http://www.typefox.io) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ide.server.concurrent;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+import org.eclipse.xtext.util.CancelIndicator;
+
+/**
+ * @author kosyakov - Initial contribution and API
+ * @since 2.11
+ */
+public class RequestCancelIndicator implements CancelIndicator, CancelChecker, Cancellable {
+	private final CompletableFuture<?> requestFuture;
+
+	public RequestCancelIndicator(CompletableFuture<?> requestFuture) {
+		this.requestFuture = requestFuture;
+	}
+
+	@Override
+	public void cancel() {
+		this.requestFuture.cancel(true);
+	}
+
+	@Override
+	public boolean isCanceled() {
+		this.checkCanceled();
```

Comment:
Was already like this, but this is pretty nasty. It'll never return true, but instead throw an error

Reply:
I don't appreciate this either. I'd like to keep the semantic cleanup separate, though.

Reply:
I'll add a comment to this impl of isCanceled rather than adjusting the semantics on the fly.